ambry-messageformat/src/main/java/com.github.ambry.messageformat/PutMessageFormatInputStream.java

```diff
@@ -38,8 +38,77 @@
  */
 public class PutMessageFormatInputStream extends MessageFormatInputStream {
 
-  public PutMessageFormatInputStream(StoreKey key, BlobProperties blobProperties, ByteBuffer userMetadata,
-      InputStream blobStream, long streamSize, BlobType blobType) throws MessageFormatException {
+  public PutMessageFormatInputStream(StoreKey key, ByteBuffer blobEncryptionKey, BlobProperties blobProperties,
+      ByteBuffer userMetadata, InputStream blobStream, long streamSize, BlobType blobType)
+      throws MessageFormatException {
+    if (MessageFormatRecord.HEADER_VERSION_TO_USE == MessageFormatRecord.Message_Header_Version_V2) {
+      createStreamWithEncryptionKeyRecord(key, blobEncryptionKey, blobProperties, userMetadata, blobStream, streamSize,
+          blobType);
+    } else {
+      createStreamWithoutEncryptionKeyRecord(key, blobProperties, userMetadata, blobStream, streamSize, blobType);
+    }
+  }
+
+  public PutMessageFormatInputStream(StoreKey key, ByteBuffer blobEncryptionKey, BlobProperties blobProperties,
+      ByteBuffer userMetadata, InputStream blobStream, long streamSize) throws MessageFormatException {
+    this(key, blobEncryptionKey, blobProperties, userMetadata, blobStream, streamSize, BlobType.DataBlob);
+  }
+
+  /**
+   * Helper method to create a stream with encryption key record. This will be the standard once all nodes in a cluster
+   * understand reading messages with encryption key record.
+   */
+  private void createStreamWithEncryptionKeyRecord(StoreKey key, ByteBuffer blobEncryptionKey,
```

Comment:
this method is not strictly w/ encryption key record(but the name says so), bcoz encryptionKeyRecord could be null. These two methods differentiates if we use HeaderV1 or HeaderV2. May be the methods can be named appropriately.

Reply:
Done.